File: Solutions.py
'''
PROBLEM 1

TIME COMPLEXITY: O(N)
SPACE COMPLEXITY: O(N)

- using iterative approach, go to the extreme left of the tree while adding the root in the stack, when root is null, pop the root from the stack and compare to prev
- prev indicates the previous element traversed in inorder traversal
- If prev is greater than the root at any point, return False otherwise return True


'''

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        prev=None
        stack=[]
        
        while(root!=None or len(stack)!=0):
            

                
            while(root!=None):
                stack.append(root)
                root=root.left
                
            root=stack.pop()
            
            if prev!=None:
                if prev.val>=root.val:
                    return False
            prev= root
            root=root.right
            
        return True

'''
PROBLEM 2

TIME COMPLEXITY: O(N)
SPACE COMPLEXITY: O(N)

- store the indexes of elements of inorder in a hashmap so that we can find the left subtree and right subtree indexes in O(1)
- traverse the preorder array, create a node for it and find the left and right subtree indexes for that node
- pass the indexes of left and right subtree to recursion to do the job


'''

class Solution:
    idx=0
    dicti={}
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if len(preorder)==0: return None
        root=preorder[0]
        
        for i in range(len(inorder)):
            self.dicti[inorder[i]]=i
        
        return self.helper(preorder,inorder, 0, len(inorder)-1)
    
    def helper(self, preorder: List[int],inorder, start, end):
        #base case
        if start>end: return None
        
        #logic
        rootIndex= self.dicti[preorder[self.idx]]
        node=TreeNode(preorder[self.idx])
        self.idx+=1
        
        node.left= self.helper(preorder, inorder, start, rootIndex-1)
        node.right= self.helper(preorder, inorder, rootIndex+1, end)
        
        return node
        
        
        
        

File: Solutions.java
//Validate BST

// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

 // LIMIT APPROACH
 class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root, null, null);
    }

    private boolean helper(TreeNode root, Integer min, Integer max)
    {
        //base
        if(root == null) return true;
        if(min!=null && root.val <= min) return false;
        if(max!=null && root.val >= max)  return false;

        //logic 
        /*boolean left = helper(root.left, min, root.val);
        boolean right = helper(root.right, root.val, max);    

        return left && right;  */
        return helper(root.left, min, root.val) && helper(root.right, root.val, max);
    }
}



//Consrtuct BT from Inorder and Preorder traversal

//Brute Force Approach 
// Time Complexity : O(n^2)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {

        if(preorder.length == 0) return null;

        int rootVal = preorder[0];
        int idx = -1;
        TreeNode root = new TreeNode(rootVal);
        

        for(int i=0;i<inorder.length;i++)
        {
            if(inorder[i]==rootVal)
            {
            idx = i;
            break;
            }
        }

        int[] inLeft = Arrays.copyOfRange(inorder, 0, idx);
        int[] preLeft = Arrays.copyOfRange(preorder, 1, inLeft.length+1);
        int[] inRight = Arrays.copyOfRange(inorder, idx+1, inorder.length);
        int[] preRight = Arrays.copyOfRange(preorder, inLeft.length+1, preorder.length);

        root.left = buildTree(preLeft, inLeft);
        root.right = buildTree(preRight, inRight);

        return root;
    }
}
